platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,495 @@
+resubmitted in the next event.
+
+### Sheriff Module
+
+When misbehavior is found for a node, it is reported to the Sheriff. This module keeps track of the different types of
+misbehavior each node is accused of, and uses this information to determine whether to "shun" or "welcome" a node. It
+also sends this information to the Execution layer, so it may record misbehavior in state, if it so chooses, or publish
+misbehavior to other nodes in the network, allowing the network as a whole to observe and report dishonest or broken
+nodes.
+
+![Sheriff](sheriff-module.png)
+
+A node may "shun" another node, by refusing to talk with it via Gossip. If a node were to be shunned by all its
+gossip peers, then it has been effectively removed from the network, as it can no longer submit events that will be
+spread through the network, and will therefore not contribute to consensus. Should a malicious node attempt to attack
+its peers, if those peers discover this attack and simply shun the node, then taking no other action, the malicious
+node is prevented from causing further harm to the network.
+
+It may be that a node is misbehaving due to a bug, or environmental issue, rather than due to malicious intent. For
+example, a broken node with network trouble may attempt to create many connections, as each prior connection having
+failed for some reason. But it could also be malicious intent. Unable to tell the difference, the Sheriff may decide to
+shun the node for some time period, and then "welcome" it back by allowing it to form connections again. It is up to
+the Sheriff's algorithms to decide on the correct response to different behaviors. These algorithms are not defined
+here, but will be defined within the Sheriff's design documentation.
+
+### Public API
+
+![Consensus API](consensus-api.png)
+
+The public API of the Consensus module forms the interface boundary between Consensus and Execution.
+
+#### initialize
+
+Called by the Execution layer when it creates a new Consensus instance. This call will include any required arguments
+to fully initialized Consensus. For example, it will include any initial state corresponding to the current round
+at which Consensus should begin work (such as the correct set of judges for that particular round). Each module within
+Consensus supports some kind of initialization API that will be called. For example, each module must be initialized
+with the initial roster for the starting round.
+
+#### destroy
+
+Called by the Execution layer when it destroys an existing Consensus instance. This is particularly needed to shut down
+network connections held by Gossip, but could be used to stop executors, background tasks, etc. as necessary.
+
+#### changeRoster
+
+Called by the Execution layer when a new roster is prepared and ready for use. 
+
+#### onBehind
+
+Called by Consensus to notify Execution that the Consensus system is very far behind in processing relative to its
```

Comment:
If reconnect is the only effect of calling `onBehind`, should this be generalized to a `requestReconnect` call?  Will there ever be other conditions which might trigger a reconnect?  

In the TSS case, we might want to cause a node to reconnect without being behind.  It could be missing critical information in the state.    I'm just giving an example of a situation where we might call reconnect without being behind.

Reply:
I think there are cases where Execution discovers that it needs to reconnect and does that separately. But from the consensus perspective, it doesn't know or care about "reconnect" as a concept. It just cares about "initialize" -- that it can be started from some round. All the reconnect concepts are in execution. So it doesn't make sense for Consensus to say "you should reconnect now". Instead it says "here is a condition I find myself in" and Execution can decide what to do with that.

I don't know of any other signal from Consensus that would trigger a reconnect.